File: scripts/tools/repo-hygiene-check.test.mjs
import assert from "node:assert/strict";
import { mkdir, mkdtemp, rm, writeFile } from "node:fs/promises";
import { spawn } from "node:child_process";
import os from "node:os";
import path from "node:path";
import test from "node:test";
import { buildHygieneReport, isHighRiskPath, parseArgs, parsePorcelainStatus } from "./repo-hygiene-check.mjs";

test("parsePorcelainStatus separates tracked and untracked entries", () => {
  assert.deepEqual(parsePorcelainStatus(" M apps/api/src/server.ts\n?? docs/new.md\nR  old.ts -> scripts/new.ts\n"), [
    { code: " M", kind: "tracked", path: "apps/api/src/server.ts" },
    { code: "??", kind: "untracked", path: "docs/new.md" },
    { code: "R ", kind: "tracked", path: "scripts/new.ts" }
  ]);
});

test("isHighRiskPath detects source and manifest files", () => {
  assert.equal(isHighRiskPath("apps/api/src/server.ts"), true);
  assert.equal(isHighRiskPath("docs/runbooks/deploy.md"), true);
  assert.equal(isHighRiskPath("package.json"), true);
  assert.equal(isHighRiskPath(".planning/reports/a.md"), false);
  assert.equal(isHighRiskPath("logs/dev/api.log"), false);
});

test("buildHygieneReport reports dirty tracked and high-risk untracked files", async () => {
  const dir = await mkdtemp(path.join(os.tmpdir(), "home-ktv-hygiene-"));

  try {
    await git(dir, ["init"]);
    await git(dir, ["config", "user.email", "test@example.com"]);
    await git(dir, ["config", "user.name", "Test User"]);
    await mkdir(path.join(dir, "apps", "api"), { recursive: true });
    await writeFile(path.join(dir, "README.md"), "initial\n");
    await writeFile(path.join(dir, "apps", "api", "server.ts"), "tracked\n");
    await git(dir, ["add", "."]);
    await git(dir, ["commit", "-m", "init"]);
    await writeFile(path.join(dir, "apps", "api", "server.ts"), "changed\n");
    await mkdir(path.join(dir, "docs"), { recursive: true });
    await writeFile(path.join(dir, "docs", "new.md"), "new\n");
    await mkdir(path.join(dir, "logs"), { recursive: true });

    const report = await buildHygieneReport({ rootDir: dir });

    assert.equal(report.summary.trackedDirty, 1);
    assert.equal(report.summary.highRiskUntracked, 1);
    assert.equal(report.trackedDirty[0]?.path, "apps/api/server.ts");
    assert.equal(report.untrackedHighRisk[0]?.path, "docs/new.md");
    assert.deepEqual(report.runtimePaths, ["logs"]);
  } finally {
    await rm(dir, { force: true, recursive: true });
  }
});

test("parseArgs supports json and fail-on-dirty", () => {
  assert.deepEqual(parseArgs(["--", "--json", "--fail-on-dirty", "--root", "/repo"]), {
    failOnDirty: true,
    help: false,
    json: true,
    rootDir: "/repo"
  });
});

function git(cwd, args) {
  return new Promise((resolve, reject) => {
    const child = spawn("git", args, { cwd });
    let stderr = "";
    child.stderr.on("data", (chunk) => {
      stderr += chunk;
    });
    child.on("close", (code) => {
      if (code === 0) {
        resolve();
      } else {
        reject(new Error(stderr || `git ${args.join(" ")} failed`));
      }
    });
  });
}

File: scripts/repo_hygiene_check_test.py
#!/usr/bin/env python3

from __future__ import annotations

import unittest

from repo_hygiene_check import is_high_risk_path, normalize_status_path, parse_args, parse_branch_summary, parse_porcelain_status


class RepoHygieneCheckTest(unittest.TestCase):
    def test_parse_porcelain_status_splits_tracked_and_untracked(self) -> None:
        entries = parse_porcelain_status(" M README.md\n?? scripts/new_tool.py\nR  old.py -> new.py\n")

        self.assertEqual(entries[0].code, " M")
        self.assertEqual(entries[0].kind, "tracked")
        self.assertEqual(entries[0].path, "README.md")
        self.assertEqual(entries[1].kind, "untracked")
        self.assertEqual(entries[1].path, "scripts/new_tool.py")
        self.assertEqual(entries[2].path, "new.py")

    def test_high_risk_paths_match_source_and_config_files(self) -> None:
        self.assertTrue(is_high_risk_path("apps/api/src/index.ts"))
        self.assertTrue(is_high_risk_path("scripts/repo_hygiene_check.py"))
        self.assertTrue(is_high_risk_path("package.json"))
        self.assertFalse(is_high_risk_path("runtime/output.json"))
        self.assertFalse(is_high_risk_path("logs/dev.log"))

    def test_parse_branch_summary_strips_git_prefix(self) -> None:
        self.assertEqual(parse_branch_summary("## main...origin/main [ahead 1]\n"), "main...origin/main [ahead 1]")
        self.assertEqual(parse_branch_summary(""), "")

    def test_normalize_status_path_uses_rename_target(self) -> None:
        self.assertEqual(normalize_status_path("old.py -> new.py"), "new.py")

    def test_parse_args_ignores_pnpm_forwarded_separator(self) -> None:
        args = parse_args(["--", "--fail-on-dirty"])

        self.assertTrue(args.fail_on_dirty)


if __name__ == "__main__":
    unittest.main()
